q2_metadata/normalization/_norm_errors.py

```diff
@@ -0,0 +1,204 @@
+        if len(non_accepted_set):
+            return 'inacceptable validations: %s' % ', '.join(non_accepted_set)
+        possible_validations_2 = possible_validations['force_to_blank_if']
+        for rule_2, rule_value_2 in self.value['force_to_blank_if'].items():
+            if not isinstance(rule_value_2, list):
+                return 'variables not in a list for condition: %s' % rule_2
+            if rule_2 not in possible_validations_2:
+                return 'inacceptable condition: %s' % rule_2
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class NormalizationError(Error):
+    """Exception raised for errors in the
+    formatting of the "validation values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'normalization'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, dict):
+            return 'is not a dictionary'
+        possible_normalizations = {'maximum', 'mininum', 'gated_value'}
+        rule_value_set = set(self.value)
+        impossible = sorted(rule_value_set ^ (rule_value_set & possible_normalizations))
+        if len(impossible):
+            return 'impossible normalization terms: %s' % ', '.join(impossible)
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class BlankError(Error):
+    """Exception raised for errors in the
+    formatting of the "blank values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'blank'
+        self.variable = variable
+        self.value = value
+        self.allowed_values = {
+            'not applicable', 'not collected',
+            'not provided', 'restricted access'
+        }
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, str):
+            return 'is not a string'
+        if self.value not in self.allowed_values:
+            return 'not in controlled vocabulary: %s' % self.value
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class MissingError(BlankError):
+    """Exception raised for errors in the
+    formatting of the "missing values" rules.
+    """
+    def __init__(self, variable, value):
+        BlankError.__init__(self, variable, value)
+        self.rule = 'missing'
+
+
+class FormatError(BlankError):
+    """Exception raised for errors in the
+    formatting of the "format values" rules.
+    """
+    def __init__(self, variable, value):
+        BlankError.__init__(self, variable, value)
+        self.rule = 'format'
+        self.allowed_values = {
```

Comment:
same

Reply:
it is that bad?

Reply:
user-defined exceptions have been removed